Use a load instruction to access a context field [v2]

mmisono opened the pull request:
This is a follow up of #988.

The previous approach is recursively traversing the AST to see if the Node accesses a context field. But it turns out that this approach has several problems. This PR, instead, introduces Type::ctx to propagate context information in a top-down way (idea by ajor). Type::ctx is basically the same as Type::cast. codegen emit load instructions in FieldAccess and ArrayAccess if the type is Type::ctx.

Please see each commit message for the details. I think all the previously discussed points are solved. 

One thing not previously discussed is the availability of  "ctx" in tracepoint. In this PR, "ctx" is concealed in tracepoint for simplicity. 

Closes #945

fbs commented on tests/semantic_analyser.cpp:
```diff
@@ -1436,6 +1436,71 @@ TEST(semantic_analyser, builtin_args)
 }
 
+TEST(semantic_analyser, type_ctx)
+{
```
would be good to test some invalid constructs here too

mmisono replied:
I added https://github.com/iovisor/bpftrace/pull/1104/commits/56184dd61acce7722cf44536c9a06f9de0d4f6f4#diff-841d87c729a5da4754074bc000917a99

mmisono commented on src/ast/codegen_llvm.cpp:
```diff
@@ -232,9 +232,9 @@ void CodegenLLVM::visit(Builtin &builtin)
+  else if (builtin.ident == "args" || builtin.ident == "ctx")
   {
-    expr_ = ctx_;
+    expr_ = b_.CreatePtrToInt(ctx_, b_.getInt64Ty());
```
`ctx_` is used:

- As an argument of helper functions
  - [CreateGetStackID()](https://github.com/iovisor/bpftrace/blob/c532159263d5286ca1ca8120c85de48c1f39ac75/src/ast/irbuilderbpf.cpp#L625), [CreatePerfEventOutput()](https://github.com/iovisor/bpftrace/blob/c532159263d5286ca1ca8120c85de48c1f39ac75/src/ast/irbuilderbpf.cpp#L667), [CreateOverrideReturn()](https://github.com/iovisor/bpftrace/blob/c532159263d5286ca1ca8120c85de48c1f39ac75/src/ast/irbuilderbpf.cpp#L696)
  - The function takes a Int8PtrTy value
- To read registers
  - ["arg"](https://github.com/iovisor/bpftrace/blob/c532159263d5286ca1ca8120c85de48c1f39ac75/src/ast/codegen_llvm.cpp#L512), ["sarg"](https://github.com/iovisor/bpftrace/blob/c532159263d5286ca1ca8120c85de48c1f39ac75/src/ast/codegen_llvm.cpp#L214), ["reg"](https://github.com/iovisor/bpftrace/blob/c532159263d5286ca1ca8120c85de48c1f39ac75/src/ast/codegen_llvm.cpp#L620), [CreateUSDTReadArgument()](https://github.com/iovisor/bpftrace/blob/c532159263d5286ca1ca8120c85de48c1f39ac75/src/ast/irbuilderbpf.cpp#L351)
  - the address is calculated by CreateGEP(), then a value is loaded
  - e.g., https://github.com/iovisor/bpftrace/blob/c532159263d5286ca1ca8120c85de48c1f39ac75/src/ast/codegen_llvm.cpp#L618-L621

The casting makes sense to me. It will make code more maintainable since bpftrace treats most variables as an integer. I'll work on it.

mmisono replied:
I made the changes (https://github.com/mmisono/bpftrace/commit/86c657ccaa5773d7cecd397c8fff286ff2fa19ee), but this caused many codegen changes (18 test failed) and would conflict #1105. So I'll keep the initial code in this PR.

mmisono commented:
mm... I'm not sure why tests fail for Static LLVM 5 Release. It works on my machine.

```
[ RUN      ] codegen.builtin_ctx_field
libclang: crash detected during parsing: {
  'source_filename' : 'definitions.h'
  'command_line_args' : ['clang', '-isystem', '/usr/local/include', '-isystem', '/bpftrace/include', '-isystem', '/usr/include'],
  'unsaved_files' : [('definitions.h', '...', 87), ('/bpftrace/include/__stddef_max_align_t.h', '...', 1771), ('/bpftrace/include/float.h', '...', 5192), ('/bpftrace/include/limits.h', '...', 3735), ('/bpftrace/include/stdarg.h', '...', 2025), ('/bpftrace/include/stddef.h', '...', 4499), ('/bpftrace/include/stdint.h', '...', 23390), ('/bpftrace/include/clang_workarounds.h', '...', 921)],
  'options' : 1,
}
stdin:1:101-113: ERROR: Unknown struct/union: 'struct x'
struct c {char c} struct x { long a; short b[4]; struct c c; struct c *d; char e[4]}kprobe:f { $x = (struct x*)ctx; @a = $x->a; @b = $x->b[0]; @c = $x->c.c; @d = $x->d->c; @e = $x->e;}
                                                                                                    ~~~~~~~~~~~~
/home/travis/build/iovisor/bpftrace/tests/codegen/common.h:45: Failure
Expected equality of these values:
  semantics.analyse()
    Which is: 1
  0
[  FAILED  ] codegen.builtin_ctx_field (41 ms)
```

ajor commented:
libclang is broken in the LLVM 5 static builds so in CI we don't run any tests which need it https://github.com/iovisor/bpftrace/blob/145041ec75eae1e58fcd3f784964303fb75c16d5/.travis.yml#L31

mmisono commented:
I see. Excluded the test in CI.

fbs commented on tests/codegen/builtin_arg.cpp:
```diff
@@ -14,14 +14,14 @@ declare i64 @llvm.bpf.pseudo(i64, i64) #0
 declare void @llvm.lifetime.start.p0i8(i64, i8* nocapture) #1
 
-define i64 @"kprobe:f"(i8* nocapture readonly) local_unnamed_addr section "s_kprobe:f_1" {
+define i64 @"kprobe:f"(i8*) local_unnamed_addr section "s_kprobe:f_1" {
```
Does the issue still exist if this changes from i8* to i64*? Marking it volatile feels a bit hacky.

mmisono replied:
Yes, I tried it but the result was the same. I also tried to use #1105, but it didn't help.

bcc uses volatile load for some register accesses (iovisor/bcc#1212). I can reproduce this error using bcc, so I created the issue iovisor/bcc#2745.

mmisono replied:
As of volatile context access, yonghong-song said:

> Using the volatile to workaround the issue is the right way to do at this point. The issue is fixed in llvm trunk (llvm 11.0): llvm/llvm-project@d96c1bb#diff-6e02af2c61254c488f583a59982eb684
> https://github.com/iovisor/bcc/issues/2745#issuecomment-583757684

llvm 11.0 just disables ReduceLoadWidth optimization.

ajor commented on src/types.h:
```diff
@@ -38,6 +38,7 @@ enum class Type
   inet,
   stack_mode,
   array,
+  ctx,
```
I think a comment above each type explaining what they're for would be useful going forward. No need to do the others but could you add one for "ctx" please?

mmisono commented on src/ast/semantic_analyser.cpp:
```diff
@@ -944,7 +971,7 @@ void SemanticAnalyser::visit(ArrayAccess &arr)
 
   if (is_final_pass()) {
-    if (type.type != Type::array)
+    if (type.type != Type::array && type.type != Type::ctx)
```
oops.. `arr.expr->type.elem_type` is `Type::none` for non-array types, so we can check this by

```c
if (!(type.type == Type::array || (type.type == Type::ctx && type.elem_type != Type::none)))
```

Is this approach OK? Alternatively, we can introduce another type (`Type::array_ctx`)  but I guess it gives little benefit considering the amount of code added.

ajor replied:
Yeah seems reasonable
Do we even need to check `type.type` here or is `type.elem_type != Type::none` enough?

mmisono commented on docs/reference_guide.md:
````diff
@@ -1658,6 +1659,19 @@ Tracing block I/O sizes > 0 bytes
 [512K, 1M)             1 |                                                    |
 ```
 
+## 10. `ctx`: BPF program context
````
I don't have a strong opinion, but there is a possibility that new fields will be added to a context struct and isn't it supporting every these field with new builtin just cumbersome? we can access these fields using `ctx` without updating bpftrace.

Related: #945

fbs replied:
No strong opinion here either, it is a helpful for the "power user" but we can always add it in to the docs later

mmisono commented on src/ast/codegen_llvm.cpp:
```diff
@@ -1197,20 +1210,42 @@ void CodegenLLVM::visit(FieldAccess &acc)
+                         b_.CreateIntToPtr(src, field_ty->getPointerTo()),
+                         field.type.size,
+                         1,
+                         true);
```
At first, I thought so too, but it turned out that the bpf_map_update_elem only accepts the [pointer in the stack](https://github.com/torvalds/linux/blob/bb6d3fb354c5ee8d6bde2d576eb7220ea09862b9/kernel/bpf/helpers.c#L56) (The definition of `ARG_PTR_TO_MAP_VALUE` is [here](https://github.com/torvalds/linux/blob/bb6d3fb354c5ee8d6bde2d576eb7220ea09862b9/include/linux/bpf.h#L217).) I'll comment on it in the code.

```
% sudo ./src/bpftrace -e 'tracepoint:block:block_rq_issue { @ = args->rwbs; }' -v
BTF: using data from /sys/kernel/btf/vmlinux
Attaching 1 probe...

Error log:
0: (bf) r3 = r1
1: (b7) r1 = 0
2: (7b) *(u64 *)(r10 -8) = r1
last_idx 2 first_idx 0
regs=2 stack=0 before 1: (b7) r1 = 0
3: (18) r1 = 0xffff970c59cab800
5: (07) r3 += 32
6: (bf) r2 = r10
7: (07) r2 += -8
8: (b7) r4 = 0
9: (85) call bpf_map_update_elem#2
R3 type=ctx expected=fp
processed 9 insns (limit 1000000) max_states_per_insn 0 total_states 0 peak_states 0 mark_read 0

Error loading program: tracepoint:block:block_rq_issue
```

(By the way, `printf("%s\n", args->rwbs);` is allowed when doing `expr_ = src` because [it accepts arbitrary memory location](https://github.com/torvalds/linux/blob/bb6d3fb354c5ee8d6bde2d576eb7220ea09862b9/kernel/trace/bpf_trace.c#L618). My guess is because bpf_map_update_elem (and other map functions) can be usable in any bpf program type it accepts only `ARG_PTR_TO_MAP_VALUE`  to prevent kernel data leakage, though as of the tracing BPF program it is pointless since it can read arbitrary memory.)

mmisono commented on src/ast/semantic_analyser.cpp:
```diff
@@ -1342,6 +1390,10 @@ void SemanticAnalyser::visit(AssignMapStatement &assignment)
   }
+  else if (type == Type::ctx)
+  {
+    error("Storing context in the map is not supported", assignment.loc);
```
The verifier rejects this and I found the error message from the verifier is ambiguous, so generates an error here. I'll comment on it here. #978

mmisono commented on src/ast/codegen_llvm.cpp:
```diff
@@ -616,10 +613,12 @@ void CodegenLLVM::visit(Call &call)
       abort();
     }
 
+    // LLVM 7.0 <= does not have CreatLoad(*Ty, *Ptr, isVolatile, Name)
```
```suggestion
    // LLVM 7.0 <= does not have CreateLoad(*Ty, *Ptr, isVolatile, Name)
```

mmisono commented:
@ajor 
I updated the code based on the above comments. Please take a look when you have time.
I'll rebase & squash commits and remove the commit of ctx documentation after the review is done.

fbs commented:
@mmisono  This looks good to me. Unless any objections pop up I want to merge this in [12-36]hours to unblock the other PRs. We still have plenty of time to fix any bugs that come up before the next release anyway :)

Can you fix the conflict and history?

mmisono commented:
@fbs 
I rebased & resolved the conflicts. Thanks very much for waiting for this!

ajor commented on src/ast/semantic_analyser.cpp:
```diff
@@ -1360,6 +1409,11 @@ void SemanticAnalyser::visit(AssignMapStatement &assignment)
+  else if (type == Type::ctx)
+  {
+    // bpf_map_update_elem only accepts a pointer to a element in the stack
+    error("Storing context in the map is not supported", assignment.loc);
```
So it's not something that can ever work with BPF? The current wording makes it sound like it's just an unimplemented feature of bpftrace.
Maybe something like "Can not assign context to a map" would be better

mmisono commented:
@ajor
I updated the comment and added the check for `@x = ctx`. Thanks for pointing it out.
https://github.com/iovisor/bpftrace/pull/1104/commits/368a64ff8d237834aedb0e1a6252b5426717b222#diff-286594496696c8553c9f12fcabfb5f38R882-R886

> Do we even need to check type.type here or is type.elem_type != Type::none enough?

I believe checking `type.elem_type != Type::none` is enough to distinguish array types, but I keep the check of `type.type` so that not assigning type.elem_type accidentally in the future.

ajor commented on src/ast/semantic_analyser.cpp:
```diff
@@ -1299,15 +1348,18 @@ void SemanticAnalyser::visit(Cast &cast)
 
     auto rhs = cast.expr->type.type;
-    if (! (rhs == Type::integer || rhs == Type::cast)) {
+    if (!(rhs == Type::integer || rhs == Type::cast || rhs == Type::ctx))
```
Is there any case where casting a `Type::ctx` is useful?

mmisono commented:
@ajor 
This is supported to access tracepoint's `__data_loc`.

> Note that casting Type::ctx to Type::integer is supported because it may
be used for some workarounds (e.g., #999). In this case, the context
information will be lost.
> https://github.com/iovisor/bpftrace/pull/1104/commits/0e736dca74d43f230d3c9955748d23bf6bbc5136

That is, we can access a `__data_loc` field like the following.

```
tracepoint:sched:sched_process_exec {
  $data = (struct data*)args;
  $offset = (uint64)($data->filename & 0xFFFF);
  $length = (uint64)($data->filename >> 16);
  $base = (uint64)args;
  printf("filename=%s\n", str($base + $offset, $length));
}
```

In this case, `str()` does `bpf_probe_read_str()`, which accepts arbitrary memory location, so the above code is OK. We need to cast a value to `uint64` to perform an arithmetic operation. Currently, I don't come up with any other use case we need to cast `args`, so I guess we can remove this casting support once #770 is completeded.

mmisono commented:
I'll add the comment about it.
